hooks/hookexecution/enricher.go

```diff
@@ -0,0 +1,184 @@
+	bidRequest *openrtb2.BidRequest,
+	account *config.Account,
+) (json.RawMessage, error) {
+	if len(stageOutcomes) == 0 {
+		return nil, nil
+	}
+
+	trace, isDebugEnabled := getDebugContext(bidRequest, account)
+	modulesOutcome := getModulesOutcome(stageOutcomes, trace, isDebugEnabled)
+	if modulesOutcome == nil {
+		return nil, nil
+	}
+
+	return json.Marshal(modulesOutcome)
+}
+
+func getDebugContext(bidRequest *openrtb2.BidRequest, account *config.Account) (trace, bool) {
+	var traceLevel string
+	var isDebugEnabled bool
+
+	if bidRequest != nil {
+		traceLevel, _ = jsonparser.GetString(bidRequest.Ext, "prebid", "trace")
+		isDebug, _ := jsonparser.GetBoolean(bidRequest.Ext, "prebid", "debug")
+		isDebugEnabled = bidRequest.Test == 1 || isDebug
+		if account != nil {
+			isDebugEnabled = isDebugEnabled && account.DebugAllow
+		}
+	}
+
+	return trace(traceLevel), isDebugEnabled
+}
+
+func getModulesOutcome(stageOutcomes []StageOutcome, trace trace, isDebugEnabled bool) *ModulesOutcome {
+	var modulesOutcome ModulesOutcome
+	stages := make(map[string]Stage)
+	stageNames := make([]string, 0)
+
+	for _, stageOutcome := range stageOutcomes {
+		if len(stageOutcome.Groups) == 0 {
+			continue
+		}
+
+		prepareModulesOutcome(&modulesOutcome, stageOutcome.Groups, trace, isDebugEnabled)
+		if !trace.isBasicOrHigher() {
+			continue
+		}
+
+		stage, ok := stages[stageOutcome.Stage]
+		if !ok {
+			stageNames = append(stageNames, stageOutcome.Stage)
+			stage = Stage{
+				Stage:    stageOutcome.Stage,
+				Outcomes: []StageOutcome{},
+			}
+		}
+
+		stage.Outcomes = append(stage.Outcomes, stageOutcome)
+		if stageOutcome.ExecutionTimeMillis > stage.ExecutionTimeMillis {
+			stage.ExecutionTimeMillis = stageOutcome.ExecutionTimeMillis
+		}
+
+		stages[stageOutcome.Stage] = stage
+	}
+
+	if modulesOutcome.Errors == nil && modulesOutcome.Warnings == nil && len(stages) == 0 {
+		return nil
+	}
+
+	if len(stages) > 0 {
+		modulesOutcome.Trace = &TraceOutcome{}
+		modulesOutcome.Trace.Stages = make([]Stage, 0, len(stages))
+
+		// iterate through slice of names to keep order of stages
+		for _, stage := range stageNames {
+			modulesOutcome.Trace.ExecutionTimeMillis += stages[stage].ExecutionTimeMillis
+			modulesOutcome.Trace.Stages = append(modulesOutcome.Trace.Stages, stages[stage])
+		}
+	}
+
+	return &modulesOutcome
+}
+
+func prepareModulesOutcome(modulesOutcome *ModulesOutcome, groups []GroupOutcome, trace trace, isDebugEnabled bool) {
+	for _, group := range groups {
+		for i, hookOutcome := range group.InvocationResults {
+			if !trace.isVerbose() {
+				group.InvocationResults[i].DebugMessages = nil
+				group.InvocationResults[i].AnalyticsTags = hookanalytics.Analytics{}
+			}
+
+			if !isDebugEnabled {
+				continue
+			}
+
+			modulesOutcome.Errors = fillMessages(modulesOutcome.Errors, hookOutcome.Errors, hookOutcome.HookID)
+			modulesOutcome.Warnings = fillMessages(modulesOutcome.Warnings, hookOutcome.Warnings, hookOutcome.HookID)
```

Comment:
It might be clearer to wrap the errors and warnings with the isDebugEnabled check:

```
if isDebugEnabled {
  modulesOutcome.Errors = fillMessages(modulesOutcome.Errors, hookOutcome.Errors, hookOutcome.HookID)
  modulesOutcome.Warnings = fillMessages(modulesOutcome.Warnings, hookOutcome.Warnings, hookOutcome.HookID)
}
```